File: src/python/xerxes/security/sandbox.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Sandbox configuration and routing.

Defines the data model and the host/sandbox routing decision logic used
by the tool dispatcher. The actual isolation mechanism is supplied by a
``SandboxBackend`` implementation (Docker, subprocess, Modal, SSH,
Singularity, Daytona). Threat model: protect the host from tool calls
the model may make on its own initiative (file writes, shell commands,
Python eval) by running designated tools inside an isolated environment
with controlled mounts, env, network, memory and timeout."""

from __future__ import annotations

import logging
import typing as tp
from dataclasses import dataclass, field
from enum import Enum

logger = logging.getLogger(__name__)


class SandboxMode(Enum):
    """Enforcement level for the sandbox router.

    Values:
        OFF: sandboxing disabled; every tool runs on the host.
        WARN: log a warning when a sandboxed tool runs, but still run on
            host. Useful for dry-runs and migration.
        STRICT: refuse to run a sandboxed tool on the host — route it to
            the configured backend or raise.
    """

    OFF = "off"
    WARN = "warn"
    STRICT = "strict"


@dataclass
class SandboxBackendConfig:
    """Backend-specific knobs (image, mounts, env).

    Attributes:
        image: container image identifier (interpreted by the backend).
        mount_paths: ``{host_path: container_path}`` bind mounts.
        mount_readonly: whether mounts are read-only.
        env_vars: environment variables forwarded into the sandbox.
        extra_args: backend-specific overrides; unrecognised keys are ignored.
    """

    image: str = "python:3.12-slim"
    mount_paths: dict[str, str] = field(default_factory=dict)
    mount_readonly: bool = True
    env_vars: dict[str, str] = field(default_factory=dict)
    extra_args: dict[str, tp.Any] = field(default_factory=dict)


@dataclass
class SandboxConfig:
    """Sandbox policy + resource limits applied across all backends.

    Attributes:
        mode: overall enforcement level (see :class:`SandboxMode`).
        sandboxed_tools: tools that must run inside the sandbox under STRICT.
        elevated_tools: tools explicitly allowed to bypass the sandbox.
        sandbox_timeout: per-call wall-clock limit in seconds.
        sandbox_memory_limit_mb: per-call memory cap in MiB.
        sandbox_network_access: if False, network is disabled in the sandbox.
        working_directory: host path used as the sandbox CWD/mount.
        backend_type: backend name to instantiate (``docker``, ``subprocess``...).
        backend_config: backend-specific configuration.
    """

    mode: SandboxMode = SandboxMode.OFF
    sandboxed_tools: set[str] = field(default_factory=set)
    elevated_tools: set[str] = field(default_factory=set)
    sandbox_timeout: float = 30.0
    sandbox_memory_limit_mb: int = 512
    sandbox_network_access: bool = False
    working_directory: str | None = None
    backend_type: str | None = None
    backend_config: SandboxBackendConfig = field(default_factory=SandboxBackendConfig)


class ExecutionContext(Enum):
    """Which execution context a tool call landed in."""

    HOST = "host"
    SANDBOX = "sandbox"


@dataclass
class ExecutionDecision:
    """Routing decision for a single tool call.

    Attributes:
        context: where the tool will actually run.
        tool_name: tool the decision applies to.
        reason: short human-readable explanation (mode, allow-list, etc.).
    """

    context: ExecutionContext
    tool_name: str
    reason: str


class SandboxExecutionUnavailableError(RuntimeError):
    """Raised when a sandboxed tool is requested but no backend is wired."""

    def __init__(self, tool_name: str) -> None:
        """Record the offending tool and format a clear error message."""

        self.tool_name = tool_name
        super().__init__(f"Tool '{tool_name}' requires sandbox execution, but no sandbox backend is configured")


class SandboxBackend(tp.Protocol):
    """Structural interface every sandbox backend implementation honours."""

    def execute(self, tool_name: str, func: tp.Callable, arguments: dict) -> tp.Any:
        """Run ``func(**arguments)`` inside the sandbox and return its result."""
        ...

    def is_available(self) -> bool:
        """Return True if the backend's underlying runtime is reachable."""
        ...

    def get_capabilities(self) -> dict[str, tp.Any]:
        """Describe what isolation, limits and image the backend provides."""
        ...


class SandboxRouter:
    """Decide and (optionally) execute tool calls under the active policy.

    The router consults ``SandboxConfig`` to choose a context. Elevated
    tools always run on the host; in OFF mode everything does; in WARN
    mode sandboxed tools log a warning but still execute on host; in
    STRICT mode they are routed to the backend. If STRICT routes a call
    but no backend is attached, :class:`SandboxExecutionUnavailableError`
    is raised."""

    def __init__(self, config: SandboxConfig | None = None, backend: SandboxBackend | None = None) -> None:
        """Bind the router to a configuration and (optionally) a backend."""

        self.config = config or SandboxConfig()
        self.backend = backend

    def decide(self, tool_name: str) -> ExecutionDecision:
        """Return the :class:`ExecutionDecision` for ``tool_name`` under current config."""

        if tool_name in self.config.elevated_tools:
            return ExecutionDecision(
                context=ExecutionContext.HOST,
                tool_name=tool_name,
                reason="Tool is marked as elevated",
            )

        if self.config.mode == SandboxMode.OFF:
            return ExecutionDecision(
                context=ExecutionContext.HOST,
                tool_name=tool_name,
                reason="Sandbox mode is off",
            )

        if tool_name in self.config.sandboxed_tools:
            if self.config.mode == SandboxMode.WARN:
                logger.warning("Tool '%s' would run in sandbox (mode=warn, running on host)", tool_name)
                return ExecutionDecision(
                    context=ExecutionContext.HOST,
                    tool_name=tool_name,
                    reason="Warn mode advisory: tool would run in sandbox, executing on host",
                )
            elif self.config.mode == SandboxMode.STRICT:
                return ExecutionDecision(
                    context=ExecutionContext.SANDBOX,
                    tool_name=tool_name,
                    reason="Strict sandbox enforcement",
                )

        return ExecutionDecision(
            context=ExecutionContext.HOST,
            tool_name=tool_name,
            reason="Tool not designated for sandbox",
        )

    def execute_in_sandbox(self, tool_name: str, func: tp.Callable, arguments: dict) -> tp.Any:
        """Dispatch ``func(**arguments)`` to the attached backend.

        Raises :class:`SandboxExecutionUnavailableError` if no backend is
        configured. Callers should obtain a HOST decision from :meth:`decide`
        before invoking this method when STRICT mode is not desired."""

        if self.backend is None:
            raise SandboxExecutionUnavailableError(tool_name)
        return self.backend.execute(tool_name, func, arguments)

File: xerxes/src/security/sandbox.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { XerxesError } from '../core/errors.js'
import type { ToolExecutionContext, ToolExecutor } from '../executors/toolRegistry.js'
import type { ToolCall } from '../types/toolCalls.js'
import { SubprocessSandboxBackend, SubprocessSandboxConfigurationError } from './subprocessSandbox.js'

export const SandboxMode = {
  OFF: 'off',
  WARN: 'warn',
  STRICT: 'strict',
} as const

export type SandboxMode = (typeof SandboxMode)[keyof typeof SandboxMode]

export const ExecutionContext = {
  HOST: 'host',
  SANDBOX: 'sandbox',
} as const

export type ExecutionContext = (typeof ExecutionContext)[keyof typeof ExecutionContext]

export interface SandboxBackendConfig {
  readonly envVars?: Readonly<Record<string, string>>
  /** Subprocess backend accepts `allowedCommands: string[]` and optional `maxOutputChars: number`. */
  readonly extraArgs?: Readonly<Record<string, unknown>>
  readonly image?: string
  readonly mountPaths?: Readonly<Record<string, string>>
  readonly mountReadonly?: boolean
}

export interface SandboxConfig {
  readonly backendConfig?: SandboxBackendConfig
  readonly backendType?: string
  readonly elevatedTools?: Iterable<string>
  readonly mode?: SandboxMode
  readonly sandboxMemoryLimitMb?: number
  readonly sandboxNetworkAccess?: boolean
  readonly sandboxTimeout?: number
  readonly sandboxedTools?: Iterable<string>
  readonly workingDirectory?: string
}

export interface ExecutionDecision {
  readonly context: ExecutionContext
  readonly reason: string
  readonly toolName: string
}

export interface SandboxExecutionRequest {
  readonly arguments: ToolCall['function']['arguments']
  readonly context: ToolExecutionContext
  readonly signal?: AbortSignal
  readonly toolName: string
}

/**
 * A remote/container backend owns real isolation and must expose only a
 * serializable tool request boundary. JavaScript closures are never shipped
 * into a sandbox.
 */
export interface SandboxBackend {
  execute(request: SandboxExecutionRequest): Promise<string>
  getCapabilities?(): Readonly<Record<string, unknown>>
  isAvailable?(): Promise<boolean> | boolean
}

export interface SandboxRouterOptions {
  readonly backend?: SandboxBackend
  readonly config?: SandboxConfig
  readonly onWarning?: (decision: ExecutionDecision) => void
}

export class SandboxExecutionUnavailableError extends XerxesError {
  readonly toolName: string

  constructor(toolName: string) {
    super("Tool '" + toolName + "' requires sandbox execution, but no sandbox backend is configured")
    this.toolName = toolName
  }
}

/**
 * Resolve host versus sandbox execution under the Xerxes sandbox policy.
 *
 * This port intentionally keeps routing independent from a specific Docker,
 * subprocess, or cloud runtime. A concrete backend receives typed data, not
 * an executable closure, so a host never accidentally runs sandbox-marked
 * code while serializing it.
 */
export class SandboxRouter {
  readonly backend: SandboxBackend | undefined
  readonly config: Readonly<RequiredSandboxConfig>
  private readonly onWarning: (decision: ExecutionDecision) => void

  constructor(options: SandboxRouterOptions = {}) {
    this.config = normalizeConfig(options.config)
    this.backend = options.backend ?? backendFromConfig(this.config)
    this.onWarning = options.onWarning ?? (() => undefined)
  }

  decide(toolName: string): ExecutionDecision {
    if (this.config.elevatedTools.has(toolName)) {
      return { context: ExecutionContext.HOST, toolName, reason: 'Tool is marked as elevated' }
    }
    if (this.config.mode === SandboxMode.OFF) {
      return { context: ExecutionContext.HOST, toolName, reason: 'Sandbox mode is off' }
    }
    if (this.config.sandboxedTools.has(toolName)) {
      if (this.config.mode === SandboxMode.WARN) {
        const decision = {
          context: ExecutionContext.HOST,
          toolName,
          reason: 'Warn mode advisory: tool would run in sandbox, executing on host',
        } as const
        this.onWarning(decision)
        return decision
      }
      return { context: ExecutionContext.SANDBOX, toolName, reason: 'Strict sandbox enforcement' }
    }
    return { context: ExecutionContext.HOST, toolName, reason: 'Tool not designated for sandbox' }
  }

  /** Execute a serializable request through a configured strict backend. */
  async executeInSandbox(request: SandboxExecutionRequest): Promise<string> {
    if (!this.backend) {
      throw new SandboxExecutionUnavailableError(request.toolName)
    }
    if (this.backend.isAvailable && !(await this.backend.isAvailable())) {
      throw new SandboxExecutionUnavailableError(request.toolName)
    }
    return this.backend.execute(request)
  }

  /** Route a complete tool call, using the host executor only when policy permits it. */
  async execute(
    call: ToolCall,
    context: ToolExecutionContext,
    executeHost: () => Promise<string>,
    signal?: AbortSignal,
  ): Promise<string> {
    const decision = this.decide(call.function.name)
    if (decision.context === ExecutionContext.HOST) {
      return executeHost()
    }
    return this.executeInSandbox({
      toolName: call.function.name,
      arguments: call.function.arguments,
      context,
      ...(signal ? { signal } : {}),
    })
  }
}

/** Wrap an existing executor without changing its agent-aware tool lookup semantics. */
export class SandboxedToolExecutor implements ToolExecutor {
  constructor(
    private readonly host: ToolExecutor,
    readonly router: SandboxRouter,
  ) {}

  execute(call: ToolCall, context: ToolExecutionContext, signal?: AbortSignal): Promise<string> {
    return this.router.execute(call, context, () => this.host.execute(call, context, signal), signal)
  }
}

interface RequiredSandboxConfig {
  readonly backendConfig: Readonly<{
    readonly envVars: Readonly<Record<string, string>>
    readonly extraArgs: Readonly<Record<string, unknown>>
    readonly image: string
    readonly mountPaths: Readonly<Record<string, string>>
    readonly mountReadonly: boolean
  }>
  readonly backendType: string | undefined
  readonly elevatedTools: ReadonlySet<string>
  readonly mode: SandboxMode
  readonly sandboxMemoryLimitMb: number
  readonly sandboxNetworkAccess: boolean
  readonly sandboxTimeout: number
  readonly sandboxedTools: ReadonlySet<string>
  readonly workingDirectory: string | undefined
}

function normalizeConfig(config: SandboxConfig | undefined): RequiredSandboxConfig {
  const source = config ?? {}
  const timeout = source.sandboxTimeout ?? 30
  const memory = source.sandboxMemoryLimitMb ?? 512
  if (!Number.isFinite(timeout) || timeout <= 0) {
    throw new RangeError('sandboxTimeout must be a positive number')
  }
  if (!Number.isInteger(memory) || memory < 16) {
    throw new RangeError('sandboxMemoryLimitMb must be an integer of at least 16')
  }
  const mode = source.mode ?? SandboxMode.OFF
  if (!Object.values(SandboxMode).includes(mode)) {
    throw new RangeError('Unknown sandbox mode: ' + mode)
  }
  const backend = source.backendConfig ?? {}
  return Object.freeze({
    mode,
    sandboxedTools: new Set(source.sandboxedTools ?? []),
    elevatedTools: new Set(source.elevatedTools ?? []),
    sandboxTimeout: timeout,
    sandboxMemoryLimitMb: memory,
    sandboxNetworkAccess: source.sandboxNetworkAccess ?? false,
    workingDirectory: source.workingDirectory,
    backendType: source.backendType,
    backendConfig: Object.freeze({
      image: backend.image ?? 'node:22-slim',
      mountPaths: Object.freeze({ ...(backend.mountPaths ?? {}) }),
      mountReadonly: backend.mountReadonly ?? true,
      envVars: Object.freeze({ ...(backend.envVars ?? {}) }),
      extraArgs: Object.freeze({ ...(backend.extraArgs ?? {}) }),
    }),
  })
}

function backendFromConfig(config: RequiredSandboxConfig): SandboxBackend | undefined {
  if (config.backendType === undefined) {
    return undefined
  }
  if (config.backendType !== 'subprocess') {
    throw new SubprocessSandboxConfigurationError(
      `backendType ${JSON.stringify(config.backendType)} is not available in the Bun runtime; pass an explicit backend`,
    )
  }
  const allowedCommands = requiredStringArray(config.backendConfig.extraArgs, 'allowedCommands')
  const maxOutputChars = optionalPositiveInteger(config.backendConfig.extraArgs, 'maxOutputChars')
  return new SubprocessSandboxBackend({
    allowedCommands,
    allowedTools: config.sandboxedTools,
    environment: config.backendConfig.envVars,
    maxTimeoutMs: Math.ceil(config.sandboxTimeout * 1_000),
    memoryLimitMb: config.sandboxMemoryLimitMb,
    networkAccessRequested: config.sandboxNetworkAccess,
    ...(config.workingDirectory === undefined ? {} : { workingDirectory: config.workingDirectory }),
    ...(maxOutputChars === undefined ? {} : { maxOutputChars }),
  })
}

function requiredStringArray(values: Readonly<Record<string, unknown>>, name: string): string[] {
  const value = values[name]
  if (!Array.isArray(value) || value.length === 0) {
    throw new SubprocessSandboxConfigurationError(`backendConfig.extraArgs.${name} must be a non-empty string array`)
  }
  const strings: string[] = []
  for (const item of value) {
    if (typeof item !== 'string') {
      throw new SubprocessSandboxConfigurationError(`backendConfig.extraArgs.${name} must be a non-empty string array`)
    }
    strings.push(item)
  }
  return strings
}

function optionalPositiveInteger(values: Readonly<Record<string, unknown>>, name: string): number | undefined {
  const value = values[name]
  if (value === undefined) {
    return undefined
  }
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value <= 0) {
    throw new SubprocessSandboxConfigurationError(`backendConfig.extraArgs.${name} must be a positive integer`)
  }
  return value
}
